dev/PipsPager/PipsPager.cpp

```diff
@@ -0,0 +1,571 @@
+    m_isPointerOver = false;
+    UpdateNavigationButtonVisualStates();
+}
+
+bool PipsPager::IsOutOfControlBounds(const winrt::Point& point) {
+    // This is a conservative check. It is okay to say we are
+    // out of the bounds when close to the edge to account for rounding.
+    const auto tolerance = 1.0;
+    const auto actualWidth = ActualWidth();
+    const auto actualHeight = ActualHeight();
+    return point.X < tolerance ||
+        point.X > actualWidth - tolerance ||
+        point.Y < tolerance ||
+        point.Y  > actualHeight - tolerance;
+}
+
+void PipsPager::UpdateIndividualNavigationButtonVisualState(
+    const bool hiddenOnEdgeCondition,
+    const ButtonVisibility visibility,
+    const wstring_view& visibleStateName,
+    const wstring_view& hiddenStateName,
+    const wstring_view& enabledStateName,
+    const wstring_view& disabledStateName) {
+
+    const auto ifGenerallyVisible = !hiddenOnEdgeCondition && NumberOfPages() != 0 && MaxVisualIndicators() > 0;
+    if (visibility != ButtonVisibility::Collapsed)
+    {
+        if ((visibility == ButtonVisibility::Visible || m_isPointerOver) && ifGenerallyVisible)
+        {
+            winrt::VisualStateManager::GoToState(*this, visibleStateName, false);
+            winrt::VisualStateManager::GoToState(*this, enabledStateName, false);
+        }
+        else
+        {
+            if (!ifGenerallyVisible)
+            {
+                winrt::VisualStateManager::GoToState(*this, disabledStateName, false);
+            }
+            else
+            {
+                winrt::VisualStateManager::GoToState(*this, enabledStateName, false);
+            }
+            winrt::VisualStateManager::GoToState(*this, hiddenStateName, false);
+        }
+    }
+}
+
+void PipsPager::UpdateNavigationButtonVisualStates() {
+    const int selectedPageIndex = SelectedPageIndex();
+    const int numberOfPages = NumberOfPages();
+
+    auto const ifPreviousButtonHiddenOnEdge = selectedPageIndex == 0;
+    UpdateIndividualNavigationButtonVisualState(ifPreviousButtonHiddenOnEdge, PreviousButtonVisibility(),
+        c_previousPageButtonVisibleVisualState, c_previousPageButtonHiddenVisualState,
+        c_previousPageButtonEnabledVisualState, c_previousPageButtonDisabledVisualState);
+
+    auto const ifNextButtonHiddenOnEdge = selectedPageIndex == numberOfPages - 1;
+    UpdateIndividualNavigationButtonVisualState(ifNextButtonHiddenOnEdge, NextButtonVisibility(),
+        c_nextPageButtonVisibleVisualState, c_nextPageButtonHiddenVisualState,
+        c_nextPageButtonEnabledVisualState, c_nextPageButtonDisabledVisualState);
+}
+
+void PipsPager::ScrollToCenterOfViewport(const winrt::UIElement sender, const int index)
+{
+    if (const auto scrollViewer = m_pipsPagerScrollViewer.get())
+    {
+        /* Vertical and Horizontal AligmentsRatio are not available until Win Version 1803 (sdk version 17134) */
+        /*
+        winrt::BringIntoViewOptions options;
```

Comment:
can you do a conditional check based on the presence of the API and fallback to the else case if it is not present ? See example usage here - [https://github.com/microsoft/microsoft-ui-xaml/blob/06b3e1df156a301afd1586451432f99e523ede7d/dev/dll/SharedHelpers.cpp#L129](https://github.com/microsoft/microsoft-ui-xaml/blob/06b3e1df156a301afd1586451432f99e523ede7d/dev/dll/SharedHelpers.cpp#L129) 

The StartBringIntoView is more accurate if the size of the items vary.

Reply:
yup, will do
I just want to check if the tests will pass first :)